### PR TITLE
List and pair

### PR DESCRIPTION
**Pull request by satac2**
nostd implementations of list and pair, they are by no means a perfect copy of the std implementation. However, I feel like I don't want to put too much more time into it, until we know if we are going to move forward with this approach.

### CODE REVIEW

File: api/include/opentelemetry/nostd/list.h [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    data_ = temp;
+  }
+
+  /* Returns the element at the front of the list. */
+  T front()
+  {
+    if (size_ == 0)
+    {
+      return T();
+    }
```
Comment (0x4b) [inline api/include/opentelemetry/nostd/list.h:76]: Just FYI, not everything is going to be default constructable. Consider just returning data_[0]. Requiring the caller to make sure the list isn't empty is reasonable, and since the list could contain T() they can't e.g. use it as a sentinel value and stop when they see it.

File: api/include/opentelemetry/nostd/list.h [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+    T *temp = new T[size_];
+
+    std::memcpy(temp, std::begin(init_list), init_list.size() * sizeof(T));
+
+    data_ = temp;
+  }
+
+  /* Returns the element at the front of the list. */
+  T front()
```
Comment (0x4b) [inline api/include/opentelemetry/nostd/list.h:71]: I think you want front() and back() to return a T&.

I'm not sure that nostd::list is going to work for non-copyable types (I haven't given it a hard look), but this is a place where that would be a problem.

File: api/include/opentelemetry/nostd/list.h [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace nostd
+{
+/* Non-standard implementation of the list data structure. */
+template <class T>
+class list
+{
+public:
+  typedef T value_type;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h:20]: Yes, it seems to be required the std::insert_iterator

File: api/include/opentelemetry/nostd/list.h [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    data_ = temp;
+
+    data_[0] = element;
+
+    /* Deallocate memory used by previous array. */
+    delete[] old_data;
+  };
+
+  /* Returns the last element and removes it from the list. */
+  T pop_back()
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h:191]: What do you suggest returning in the case of an empty list instead of T()?

[the remark this answers is not in the document]
Comment (0x4b) [reply api/include/opentelemetry/nostd/list.h:191]: Sorry, forget I mentioned it. It's an error to call pop_*() on an empty list.

I was too focused on the possibility of 
T *temp = new T[-1]
but you'll already have choked on accessing data_[-1]

File: api/include/opentelemetry/nostd/list.h [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+class list
+{
+public:
+  typedef T value_type;
+
+  /* Custom iterator class*/
+  class iterator
+  {
+  public:
+    typedef std::forward_iterator_tag iterator_category;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Removed.

File: api/include/opentelemetry/nostd/list.h [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    bool operator!=(const iterator &other) { return ptr_ != other.ptr_; }
+
+  private:
+    T *ptr_;
+  };
+
+  list()
+  {
+    size_ = 0;
+    data_ = NULL;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed.

File: api/include/opentelemetry/nostd/list.h [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+    return front_elem;
+  }
+
+  /* Returns the number of elements in the list. */
+  int size() { return size_; }
+
+  void operator=(const list &other_list)
+  {
+    data_ = other_list.data_;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed from pointer assignment to memcpy and added tests.

File: api/test/nostd/list_test.cc [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    counter++;
+  }
+}
+
+/* Tests that the push_front method added elements to the front of the list. */
+TEST(ListTest, PushFront)
+{
+  list<list<int>> list_list;
+  list<int> inner_list = {0, 1, 2, 3, 4};
+  list<int> mid_list   = {9, 9, 9, 9, 9};
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: Changed to different values.

File: api/include/opentelemetry/nostd/list.h [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+#include <iterator>
+#include <memory>
+
+#include "opentelemetry/nostd/utility.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace nostd
+{
+/* Non-standard implementation of the list data structure. */
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed to // style.

File: api/test/nostd/list_test.cc [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  EXPECT_EQ(int_list.pop_back(), 4);
+  EXPECT_EQ(int_list.size(), previous_size - 1);
+
+  EXPECT_EQ(int_list.back(), 3);
+}
+
+/* Tests that the pop_front method, returns and removes the front element
+ * from the list.
+ */
+TEST(ListTest, PopFront)
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: I have changed them, I am slightly concerned that they are now too wordy though.

File: api/include/opentelemetry/nostd/list.h [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    T *old_data = data_;
+
+    data_ = temp;
+
+    auto iter_pos = begin();
+    /* Iterate until the inserted element. */
+    for (int i = 0; i < insert_index; i++)
+    {
+      ++iter_pos;
+    }
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed.

File: api/include/opentelemetry/nostd/list.h [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    int remaining_elements = previous_size - counter;
+
+    /* Copy the original data that comes before the insertion. */
+    std::memcpy(temp, data_, insert_index * sizeof(T));
+    /* Copy the inserted data. */
+    std::memcpy(temp + insert_index, &element, sizeof(T));
+    /* Copy the original data that comes after the insertion. */
+    std::memcpy(temp + next_index, data_ + counter, remaining_elements * sizeof(T));
+
+    T *old_data = data_;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Has been deleted directly.

File: api/include/opentelemetry/nostd/list.h [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+   * iterator position.
+   */
+  iterator insert(iterator position, const T &element)
+  {
+    int previous_size = size_;
+    size_++;
+    T *temp = new T[size_];
+
+    /* Counter tracks the index of the passed in iterator position. */
+    int counter = 0;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Has been done.

File: api/include/opentelemetry/nostd/list.h [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+    size_ = 0;
+    data_ = NULL;
+  };
+
+  /* Constructs class with passed in element. */
+  list(T element)
+  {
+    size_ = 1;
+
+    T *data_ = new T[size_];
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Removed local data_ and added test.

File: api/include/opentelemetry/nostd/list.h (hunk 1 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed from T() to data_[0].

File: api/include/opentelemetry/nostd/list.h (hunk 2 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed for both List and for first() and second() in pair.

File: api/test/nostd/list_test.cc [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+  for (auto iter = std::begin(insert_list); iter != std::end(insert_list); ++iter)
+  {
+    back = *iter;
+  }
+
+  int counter = 1;
+  for (auto iter = std::begin(int_list); iter != std::end(int_list); ++iter)
+  {
+    EXPECT_EQ(*iter, counter);
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc:73]: I am having some trouble finding the header that contains ElementsAre

File: api/include/opentelemetry/nostd/list.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed from pointer assignment to memcpy, and added tests.

File: api/include/opentelemetry/nostd/list.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed to nullptr.

File: api/include/opentelemetry/nostd/list.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Removed.

File: api/test/nostd/list_test.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (adityashirodkar) [reply api/test/nostd/list_test.cc:73]: It should be in gmock/gmock.h

File: api/include/opentelemetry/nostd/list.h [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+class list
+{
+public:
+  typedef T value_type;
+
+  // Custom iterator class
+  class iterator
+  {
+  public:
+    iterator(T *ptr) : ptr_(ptr) {}
```
Comment (0x4b) [inline api/include/opentelemetry/nostd/list.h:27]: Should the ctor be private (and accessible to the friend class list)?

File: api/test/nostd/list_test.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (0x4b) [reply api/test/nostd/list_test.cc:73]: FYI: https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md#container-matchers

File: api/include/opentelemetry/nostd/list.h [hunk 17]
```diff
@@ -153,92 +132,97 @@ class list
...
+    if(size_ != other_list.size_){
+      return false;
+    }
+
+    if( memcmp(data_, other_list.data_, size_) == 0){
+      return true;
+    }
+    else{
+      return false;
+    }
```
Comment (0x4b) [inline api/include/opentelemetry/nostd/list.h:224]: Generally prefer to return the result of a boolean statement directly over putting it in a conditional and then returning true / false on the corresponding branches.

```suggestion
    return (memcmp(data_, other_list.data_, size_) == 0);
```

File: api/test/nostd/list_test.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc:73]: I have tried gmock/gmock.h and am getting "'ElementsAre' was not declared in this scope"

File: api/include/opentelemetry/nostd/pair.h [hunk 18]
```diff
@@ -33,11 +33,21 @@ class pair
...
-  T1 first() { return first_; }
+  bool operator==(const pair &other_pair)
+  {
+    if(first_ == other_pair.first_){
+      if(second_ == other_pair.second_){
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (0x4b) [inline api/include/opentelemetry/nostd/pair.h:44]: WDYT of

```suggestion
    return (first_ == other_pair.first_ &&
           second_ == other_pair.second_);
```
?

File: api/test/nostd/pair_test.cc [hunk 19]
```diff
@@ -14,3 +14,65 @@ TEST(PairTest, BasicConstruction)
...
+// Tests that the comparison operator returns true when pair values
+//  are the same and returns false when different.
+TEST(PairTest, ComparisonSameValuesReturnTrue){
+
+  int val_1                = 1;
+  int val_2                = 2;
+  pair<int, int> pair_1 = pair<int, int>(val_1, val_2);
+  pair<int, int> pair_2 = pair<int, int>(val_1, val_2);
+
+  EXPECT_EQ(pair_1 == pair_2, true);
```
Comment (0x4b) [inline api/test/nostd/pair_test.cc:27]: Just FYI: Since pair::operator== is defined I think you can do
  EXPECT_EQ(pair_1, pair_2)
and 
  EXPECT_NE(pair_1, pair_2)

File: api/test/nostd/list_test.cc [hunk 20]
```diff
@@ -130,3 +131,43 @@ TEST(ListTest, PopFront)
...
+  list<int> first_list = {0, 1, 2, 3, 4};
+  list<int> first_list_copy = {0, 1, 2, 3, 4};
+  
+  list<int> second_list = {5, 6, 7, 8, 9};
+  second_list = first_list;
+  second_list.pop_back();
+  second_list.pop_back();
+  second_list.pop_back();
+
+  EXPECT_EQ(first_list == first_list_copy, true);
```
Comment (0x4b) [inline api/test/nostd/list_test.cc:149]: I think you can do
  EXPECT_EQ(first_list, first_list_copy);
  EXPECT_NE(first_list, second_list);
since list::operator== is defined, but if not then
  EXPECT_TRUE(...) and EXPECT_FALSE(...) are preferable to EXPECT_EQ(foo, true/false).

File: api/test/nostd/list_test.cc [hunk 21]
```diff
@@ -130,3 +131,43 @@ TEST(ListTest, PopFront)
...
+// Tests that after using the assignment operator, changing the new 
+// doesn't change the original
+TEST(ListTest,AssignmentOperatorDoesNotChangeOriginal)
+{
+
+  list<int> first_list = {0, 1, 2, 3, 4};
+  list<int> first_list_copy = {0, 1, 2, 3, 4};
+  
+  list<int> second_list = {5, 6, 7, 8, 9};
+  second_list = first_list;
```
Comment (0x4b) [inline api/test/nostd/list_test.cc:144]: I'm assuming that the initialization of second_list isn't important, and only the subsequent assignment is. If that's true, then please initialize it directly from first_list (or if necessary, don't initialize it at declaration).

If both steps are necessary, then please add a comment about why.

File: api/include/opentelemetry/nostd/list.h [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+  }
+  
+  // Returns true if the lists are the same, false if they are not
+  bool operator==(const list &other_list)
+  {
+    if(size_ != other_list.size_){
+      return false;
+    }
+
+    if( memcmp(data_, other_list.data_, size_) == 0){
```
Comment (adityashirodkar) [inline api/include/opentelemetry/nostd/list.h:219]: Same comment as for `pair` below.

File: api/test/nostd/list_test.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (adityashirodkar) [reply api/test/nostd/list_test.cc:73]: Probably you're missing the namespace? testing::ElementsAre

[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc:73]: It looks like in order to use ElementsAre, I would also need to implement a const iterator for nostd::list. Do you think that would be valuable?

File: api/test/nostd/pair_test.cc (hunk 19 above)
Comment (satac2) [reply api/test/nostd/pair_test.cc:27]: When I try, it spits out a number of errors, does that imply that the == operator is not implemented properly?

File: api/include/opentelemetry/nostd/list.h (hunk 16 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Ctor moved to private.

File: api/include/opentelemetry/nostd/list.h (hunk 17 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Changed to removed if statements.

File: api/include/opentelemetry/nostd/pair.h (hunk 18 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/pair.h]: Removed if statements.

File: api/test/nostd/list_test.cc (hunk 20 above)
Comment (satac2) [reply api/test/nostd/list_test.cc]: Switched to EXPECT_TRUE and EXPECT_FALSE.

File: api/test/nostd/list_test.cc (hunk 21 above)
Comment (satac2) [reply api/test/nostd/list_test.cc]: Condensed to one step.

File: api/test/nostd/list_test.cc [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+// other than an int. 
+TEST(ListTest, ListAcceptsPairDatatype)
+{
+  list<pair<int, int>> pair_list;
+
+  pair_list.push_back(pair<int, int>(1, 2));
+  pair_list.push_back(pair<int, int>(2, 3));
+  pair_list.push_back(pair<int, int>(4, 5));
+
+  pair<int, int> temp_pair;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: Removed temp_pair.

File: api/test/nostd/list_test.cc [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+
+using opentelemetry::nostd::list;
+using opentelemetry::nostd::pair;
+
+// Tests that front() returns the first element and back()
+// returns the last element.
+ 
+TEST(ListTest, FrontBackReturnsProperElements)
+{
+  list<int> list_1;
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: Renamed to int_list.

File: api/test/nostd/list_test.cc [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+
+  EXPECT_EQ(pair_list.back().first(), 4);
+  EXPECT_EQ(pair_list.back().second(), 5);
+}
+
+// Tests that a list object can be initialized with an initalizer list.
+TEST(ListTest, ListConstructorAcceptsInitializerList)
+{
+
+  list<int> pair_list = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: Renamed to int_list.

File: api/test/nostd/list_test.cc [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+
+// Tests that after using the assignment operator, changing the new 
+// doesn't change the original
+TEST(ListTest,AssignmentOperatorDoesNotChangeOriginal)
+{
+
+  list<int> first_list = {0, 1, 2, 3, 4};
+  list<int> first_list_copy = {0, 1, 2, 3, 4};
+  
+  list<int> second_list = {5, 6, 7, 8, 9};
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/list_test.cc]: Changed to 

> list<int> second_list = first_list

File: api/test/nostd/pair_test.cc [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+  pair_1.second() = 7;
+
+  EXPECT_EQ(pair_1.first(), 8);
+  EXPECT_EQ(pair_1.second(), 7);
+  
+  EXPECT_EQ(pair_2.first(), val_3);
+  EXPECT_EQ(pair_2.second(), val_4);
+}
+
+
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/test/nostd/pair_test.cc:92]: I have added a test that takes containers.

File: api/include/opentelemetry/nostd/pair.h [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+  void operator=(const pair &other_pair)
+  {
+    first_  = other_pair.first_;
+    second_ = other_pair.second_;
+  }
+
+  bool operator==(const pair &other_pair)
+  {
+    if(first_ == other_pair.first_){
```
[the remark this answers is not in the document]
Comment (satac2) [reply api/include/opentelemetry/nostd/pair.h]: Simplified.

File: api/include/opentelemetry/nostd/list.h (hunk 22 above)
Comment (satac2) [reply api/include/opentelemetry/nostd/list.h]: Simplified.

File: api/test/nostd/list_test.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (0x4b) [reply api/test/nostd/list_test.cc:67]: I wouldn't really recommend extending the nostd::list interface specifically for testing, but a const iterator seems like it would be useful generally. Do it if you think it will be quick and you have some spare time.
